pkg/resources/kafka/poddisruptionbudget.go

```diff
@@ -0,0 +1,108 @@
+// Copyright © 2019 Banzai Cloud
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package kafka
+
+import (
+	"fmt"
+	"math"
+	"strconv"
+	"strings"
+
+	"github.com/banzaicloud/kafka-operator/pkg/resources/templates"
+	"github.com/banzaicloud/kafka-operator/pkg/util"
+	"github.com/go-logr/logr"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/intstr"
+
+	policyv1beta1 "k8s.io/api/policy/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func (r *Reconciler) poddisruptionbudget(log logr.Logger) (runtime.Object, error) {
+	minAvailable, err := r.computeMinAvailable(log)
+
+	if err != nil {
+		return nil, err
+
+	}
+
+	return &policyv1beta1.PodDisruptionBudget{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "PodDisruptionBudget",
+			APIVersion: "policy/v1beta1",
+		},
+		ObjectMeta: templates.ObjectMetaWithAnnotations(
+			fmt.Sprintf("%s-pdb", r.KafkaCluster.Name),
+			util.MergeLabels(LabelsForKafka(r.KafkaCluster.Name), r.KafkaCluster.Labels),
+			r.KafkaCluster.Spec.ListenersConfig.GetServiceAnnotations(),
+			r.KafkaCluster,
+		),
+		Spec: policyv1beta1.PodDisruptionBudgetSpec{
+			MinAvailable: &minAvailable,
+			Selector: &metav1.LabelSelector{
+				MatchLabels: LabelsForKafka(r.KafkaCluster.Name),
+			},
+		},
+	}, nil
+
+}
+
+// Calculate maxUnavailable as max between brokerCount - 1 (so we only allow 1 broker to be disrupted)
+// and 1 (to cover for 1 broker clusters)
+func (r *Reconciler) computeMinAvailable(log logr.Logger) (intstr.IntOrString, error) {
+
+	/*
+		budget = r.KafkaCluster.Spec.DisruptionBudget.budget (string) ->
+		- can either be %percentage or static number
+
+		Logic:
+
+		Max(1, brokers-budget) - for a static number budget
+
+		Max(1, brokers-brokers*percentage) - for a percentage budget
+
+	*/
+
+	// number of brokers in the KafkaCluster
+	brokers := len(r.KafkaCluster.Spec.Brokers)
+
+	// configured budget in the KafkaCluster
+	disruptionBudget := r.KafkaCluster.Spec.DisruptionBudget.Budget
+
+	budget := 0
+
+	// treat percentage budget
+	if strings.HasSuffix(disruptionBudget, "%") {
+		percentage, err := strconv.ParseFloat(disruptionBudget[:len(disruptionBudget)-1], 4)
+		if err != nil {
+			log.Error(err, "error occurred during parsing the disruption budget")
+			return intstr.FromInt(-1), err
+		} else {
+			budget = int(math.Floor((percentage * float64(brokers)) / 100))
```

Comment:
Since PDB allows percentage as well should just return `intstr.FromString(disruptionBudget)` here and use that in `PodDisruptionBudgetSpec`.

Reply:
Sadly no, we have to compute that since it is only supported when the scale resource is implemented in the CRD.
https://kubernetes.io/docs/tasks/run-application/configure-pdb/#arbitrary-controllers-and-selectors